lib/gringotts/gateways/mercadopago.ex

```diff
@@ -0,0 +1,272 @@
+
+  [gs]: https://github.com/aviabird/gringotts/wiki/
+  [home]: https://www.mercadopago.com
+  [example]: https://github.com/aviabird/gringotts_example
+  """
+
+  # The Base module has the (abstract) public API, and some utility
+  # implementations.  
+  @base_url "https://api.mercadopago.com"
+  use Gringotts.Gateways.Base
+  alias Gringotts.CreditCard
+  # The Adapter module provides the `validate_config/1`
+  # Add the keys that must be present in the Application config in the
+  # `required_config` list
+  use Gringotts.Adapter, required_config: [:public_key, :access_token]
+  
+  import Poison, only: [decode: 1]
+
+  alias Gringotts.{CreditCard, Response}
+
+  @doc """
+  Performs a (pre) Authorize operation.
+
+  The authorization validates the `card` details with the banking network,
+  places a hold on the transaction `amount` in the customer’s issuing bank.
+
+  mercadoapgo's `authorize` returns authorization ID(available in the `Response.id` field) :
+
+  * `capture/3` _an_ amount.
+  * `void/2` a pre-authorization.
+  ## Note
+
+  For a new customer, `customer_id` field should be ignored. Otherwise it should be provided.
+
+  ## Example
+
+  ### Authorization for new customer.
+    The following example shows how one would (pre) authorize a payment of 42 BRL on a sample `card`.
+    Ignore `customer_id`.
+      iex> amount = Money.new(42, :BRL)
+      iex> card = %Gringotts.CreditCard{first_name: "Lord", last_name: "Voldemort", number: "4509953566233704", year: 2099, month: 12, verification_code: "123", brand: "VISA"}
+      iex> opts = [email: "tommarvolo@riddle.com", order_id: 123123, payment_method_id: "visa", config: %{access_token: YOUR_ACCESS_TOKEN, public_key: YOUR_PUBLIC_KEY}]
+      iex> {:ok, auth_result} = Gringotts.authorize(Gringotts.Gateways.Mercadopago, amount, card, opts)
+      iex> auth_result.id # This is the authorization ID
+      iex> auth_result.token # This is the customer ID/token
+
+  ### Authorization for old customer.
+    The following example shows how one would (pre) authorize a payment of 42 BRL on a sample `card`.
+    Mention `customer_id`.  
+      iex> amount = Money.new(42, :BRL)
+      iex> card = %Gringotts.CreditCard{first_name: "Hermione", last_name: "Granger", number: "4509953566233704", year: 2099, month: 12, verification_code: "123", brand: "VISA"}
+      iex> opts = [email: "hermione@granger.com", order_id: 123125, customer_id: "308537342-HStv9cJCgK0dWU", payment_method_id: "visa", config: %{access_token: YOUR_ACCESS_TOKEN, public_key: YOUR_PUBLIC_KEY}]
+      iex> {:ok, auth_result} = Gringotts.authorize(Gringotts.Gateways.Mercadopago, amount, card, opts)
+      iex> auth_result.id # This is the authorization ID
+      iex> auth_result.token # This is the customer ID/token
+
+  """
+
+  @spec authorize(Money.t(), CreditCard.t(), keyword) :: {:ok | :error, Response}
+  def authorize(amount , %CreditCard{} = card, opts) do
+    if Keyword.has_key?(opts, :customer_id) do
```

Comment:
We can move this check inside `create_customer`, since `create_customer` recieves the same `opts`.
```elixir
def authorize(amount , %CreditCard{} = card, opts) do
  with {:ok, customer_id} <- create_customer(opts),
     {:ok, card_token} <- create_token(card, opts) do
    {_, value, _, _} = Money.to_integer_exp(amount)
    url_params = [access_token: opts[:config][:access_token]]
    body = authorize_params(value, card, opts, card_token, customer_id, false) |> Poison.encode!
    commit(:post, "/v1/payments", body, opts, params: url_params )
  end
end
```
> If you wish you could even write an `else` block. Here are the [`with/1` docs](https://hexdocs.pm/elixir/Kernel.SpecialForms.html#with/1).

This way we can get rid of `auth_token/4` as well as `auth/5`. Please refrain from writing so many helper functions.